src/QldbDriver.ts

```diff
@@ -156,36 +169,95 @@ export class QldbDriver {
+                        return session;
+                    } catch (e) {
+                        // An error when failing to start a new session is always retriable
+                        throw new ExecuteError(e, true, true);
+                    }
+                } else {
+                    throw new SessionPoolEmptyError()
                 }
-                //If it is ISE but not because of transaction expiry, then pick new session and retry the transaction
-                if (isInvalidSessionException(err) && !isTransactionExpiredException(err) ) {
-                    continue;
+            }
+
+        // Release semaphore and if the session is alive return it to the pool and return true
+        const releaseSession: (thisDriver: QldbDriver, session: QldbSession) => boolean = 
+            function(thisDriver: QldbDriver, session: QldbSession): boolean {
+                thisDriver._semaphore.release();
+                thisDriver._availablePermits++;
+                if (session != null && session.isAlive()) {
+                    thisDriver._sessionPool.push(session);
+                    return true;
                 } else {
-                    throw err;
+                    return false;
                 }
-            } finally {
-                if (session != null) {
-                    this._returnSessionToPool(session);
+            }
+        
+        retryConfig = (retryConfig == null) ? this._retryConfig : retryConfig;
+        let session: QldbSession;
+        let startNewSession: boolean = false;
+        for (let retryAttempt: number = 1; true; retryAttempt++) {
+            try {
+                session = null;
+                session = await getSession(this, startNewSession);
+                return await session.executeLambda(transactionLambda);
+            } catch (e) {
+                if (e instanceof ExecuteError) {
+                    if (e.isRetriable) {
+                        // Always retry on the first attempt if failure was caused by a stale session in the pool 
+                        if (retryAttempt == 1 && e.isISE) {
+                            debug("Initial session received from pool is invalid. Retrtying...");
+                            continue;
+                        }
+                        if (retryAttempt > retryConfig.getRetryLimit()) {
+                            throw e.cause;
+                        }
+
+                        const backoffFunction: BackoffFunction = retryConfig.getBackoffFunction();
+                        let backoffDelay: number = backoffFunction(retryAttempt, e.cause, e.transactionId);
+                        if (backoffDelay == null || backoffDelay < 0) {
+                            backoffDelay = 0;
+                        }
+                        await new Promise(resolve => setTimeout(resolve, backoffDelay));
+
+                        info(`A recoverable error has occured. Attempting retry #${retryAttempt}.`);
```

Comment:
Debug-level I think

Reply:
The rest of the drivers have this on the info level.

Reply:
*occurred